lib/clerk/authenticate_request.rb

```diff
@@ -0,0 +1,247 @@
+module Clerk
+    ##
+    # This class represents a service object used to determine the current request state
+    # for the current env passed based on a provided Clerk::AuthenticateContext.
+    # There is only 1 public method exposed (`resolve`) to be invoked with a env parameter.
+    class AuthenticateRequest
+        attr_reader :auth_context
+
+        ##
+        # Creates a new instance using Clerk::AuthenticateContext object.
+        def initialize(auth_context)
+            @auth_context = auth_context
+        end
+
+        ##
+        # Determines the current request state by verifying a Clerk token in headers or cookies.
+        # The possible outcomes of this method are `signed-in`, `signed-out` or `handshake` states.
+        # The return values are the same as a return value of a rack middleware `[http_status_code, headers, body]`.
+        # When used in a middleware the consumer of this service should return the return value when there is an
+        # `http_status_code` provided otherwise the should continue with the middleware chain.
+        # The headers provided in the return value is a hash of { header_key => header_value } and in the case
+        # of a `Set-Cookie` header the `header_value` used is a list of raw HTTP Set-Cookie directives.
+        def resolve(env)
+            if auth_context.header_token?
+                resolve_header_token(env)
+            else
+                resolve_cookie_token(env)
+            end
+        end
+
+        protected
+
+        def resolve_header_token(env)
+            begin
+                # malformed JWT
+                return signed_out(reason: TokenVerificationErrorReason::TOKEN_INVALID) if !sdk.decode_token(auth_context.header_token)
+            rescue JWT::DecodeError
+                # malformed JSON authorization header
+                return signed_out(reason: TokenVerificationErrorReason::TOKEN_INVALID)
+            end
+
+            begin
+                claims = verify_token(auth_context.header_token)
+                return signed_in(env, claims, auth_context.header_token) if claims
+            rescue JWT::ExpiredSignature
+                signed_out(enforce_auth: true, reason: TokenVerificationErrorReason::TOKEN_EXPIRED)
+            rescue JWT::InvalidIatError
+                signed_out(enforce_auth: true, reason: TokenVerificationErrorReason::TOKEN_NOT_ACTIVE_YET)
+            end
+
+            # Clerk.js should refresh the token and retry
+            return signed_out(enforce_auth: true)
+        end
+
+        def resolve_cookie_token(env)
+            # in cross-origin XHRs the use of Authorization header is mandatory.
+            # TODO: add reason
+            return signed_out if auth_context.cross_origin_request?
```

Comment:
❓ Shouldn't this check go to the resolve_header_token method?

Reply:
Since the condition used to execute the `resolve_header_token` flow is the existence of a `header_token` and the case described here is that we want to return a `signed_out` state if it's a cross-origin request and there is no header_token, it will never be triggered in the `resolve_header_token` flow.